matrix_sdk/src/room/joined.rs

```diff
@@ -168,14 +170,46 @@ impl Joined {
     ///
     /// ```
     pub async fn typing_notice(&self, typing: impl Into<Typing>) -> Result<()> {
-        // TODO: don't send a request if a typing notice is being sent or is already active
-        let request = TypingRequest::new(
-            self.inner.own_user_id(),
-            self.inner.room_id(),
-            typing.into(),
-        );
+        let typing = typing.into();
+
+        let send = if let Some((was_typing, mutex)) = self
+            .client
+            .typing_notice_locks
+            .get(self.inner.room_id())
+            .map(|m| m.clone())
+        {
+            let is_typing = matches!(typing, Typing::Yes(_));
+
+            if was_typing == is_typing {
+                mutex.lock().await;
+                false
+            } else {
+                true
+            }
+        } else {
+            true
+        };
+
+        if send {
+            let is_typing = matches!(typing, Typing::Yes(_));
+            let mutex = Arc::new(Mutex::new(()));
+            self.client
+                .typing_notice_locks
+                .insert(self.inner.room_id().clone(), (is_typing, mutex.clone()));
+
+            let _guard = mutex.lock().await;
+
+            let request =
+                TypingRequest::new(self.inner.own_user_id(), self.inner.room_id(), typing);
+
+            self.client.send(request, None).await?;
+
+            if let Typing::Yes(duration) = typing {
+                tokio::time::sleep(duration).await;
```

Comment:
i didn't know that sleep doesn't work in WASM. I implemented this already before you suggested the snippet. :)
Are there docs some where on what doesn't work in WASM? Maybe we should add a CI job for WASM to catch this.

Reply:
To me it seams clear to hold the lock for the duration, then checking if the `duration` has passed.

Reply:
> i didn't know that sleep doesn't work in WASM. I implemented this already before you suggested the snippet. :)
> Are there docs some where on what doesn't work in WASM? Maybe we can add a CI job for WASM to catch this.

We had CI set up to compile the WASM bot example, though it was lost in the transition from Travis to Github CI. If it's using `tokio` it's likely that it doesn't work. Spawning threads and some time based methods are common culprits that don't work.

Reply:
i see, i will change the code :+1:

Reply:
How does Fractal intend to use that method? In Weechat I have a callback that gets called every time the input changes, thus I'll want a method where I can safely call `set_typing()` and `unset_typing()` every time the callback gets executed and be done with it (note, not that I want two methods, it can be a single one with arguments). If there's nothing to do, the method turns into a NOP. 

I guess both ways accomplish this, just that it's a bit hard to sleep under WASM.

Reply:
It would be nice if the user hasn't to think about about the timeout duration. Can we just set a default timeout, without exposing it to the user?

Reply:
it would make this a much nice api.

Reply:
Sure, if you don't have a concrete value in mind, 4s from Weechat seemed to work nicely. And if someone really want s to manage this manually we can provide a dumb method on the `Client` object.

Reply:
I made it now the same way as in the snippet, but I moved the `Instant::now()` before the request to the server which allows us to remove the extra lock and it takes in account the time for the request.

I'm still not completely happy with it. Every `4s` the type notice will be deactivated and the user can't do anything about it. They obviously can reactivated, but i would prefer that the user can keep it alive.

Reply:
One solution i can think of is to allow send a new request every `TYPING_NOTICE_TIMEOUT/2`, which would make sure that the user has 2s to renew the type notice before it runes out.

Reply:
Yeah, we can certainly allow sending out a fresh notice if the existing one is about to expire, though 2s seems a bit much, 1s or 0.5 seems to be enough under the assumption that a request would take around 50-100ms?

Reply:
I set it to 1s. Now I'm fine with the API.